diff -urN elinks-20020208/configure.in aaa/elinks-20020208/configure.in
--- elinks-20020208/configure.in	Thu Jan 17 21:58:00 2002
+++ aaa/elinks-20020208/configure.in	Sun Feb 10 22:52:30 2002
@@ -138,7 +138,7 @@
 	for luadir in "" /usr /usr/local; do
 		if test "$cf_result" = no; then
 			if test ! -z "$luadir"; then
-				LIBS="-L$luadir/lib -llua -llualib -lm $LIBS_X"
+				LIBS="-L$luadir/lib -llua -llualib -lm -ldl $LIBS_X"
 				CFLAGS="$CFLAGS_X -I$luadir/include"
 			else
 				LIBS="-llua -llualib -lm $LIBS_X"
