diff -urN aaa/elinks-0.4pre10/contrib/lua/hooks.lua.in elinks-0.4pre10/contrib/lua/hooks.lua.in
--- aaa/elinks-0.4pre10/contrib/lua/hooks.lua.in	Wed Jul  3 00:38:05 2002
+++ elinks-0.4pre10/contrib/lua/hooks.lua.in	Thu Jul  4 13:07:55 2002
@@ -9,7 +9,7 @@
 --  Load configuration
 ----------------------------------------------------------------------
 
-home_dir = (getenv and getenv ("HOME")) or "/home/"..getenv("USER")
+home_dir = (getenv and getenv ("HOME")) or (getenv and "/home/"..getenv("USER"))
 dofile ("@sysconfdir@/config.lua")
 dofile (home_dir.."/.elinks/config.lua")
 
