--- elinks-0.10.0/src/mime/mime.c.orig	2004-10-14 20:50:37.000000000 +0200
+++ elinks-0.10.0/src/mime/mime.c	2004-12-31 13:26:02.000000000 +0100
@@ -177,6 +177,8 @@
 
 	ctype = parse_header(cached->head, "Content-Type", NULL);
 	if (ctype) {
+	/* treat empty Content-Type as not present (to avoid crash) */
+	if (*ctype != '\0') {
 		unsigned char *end = strchr(ctype, ';');
 		int ctypelen;
 
@@ -191,7 +193,7 @@
 		if (*ctype) {
 			return ctype;
 		}
-
+	}
 		mem_free(ctype);
 	}
 
