--- elinks.orig/src/lowlevel/home.c	Sun Oct  5 07:51:20 2003
+++ elinks/src/lowlevel/home.c	Sun Oct  5 12:25:50 2003
@@ -109,7 +109,7 @@
 get_home(void)
 {
 	unsigned char *home_elinks;
-	unsigned char *envhome = getenv("HOME");
+	unsigned char *envhome = getenv("HOME_ETC") ? getenv("HOME_ETC") : getenv("HOME");
 	unsigned char *home = envhome ? stracpy(envhome)
 				      : elinks_dirname(path_to_exe);
 
--- elinks/doc/man/man1/elinks.1.in.orig	Thu Oct 23 13:07:37 2003
+++ elinks/doc/man/man1/elinks.1.in	Thu Oct 23 13:12:03 2003
@@ -127,6 +127,10 @@
 The location of the directory containing configuration files.
 If not set the default is \fI~/.elinks/\fR.
 .TP
+\fB HOME_ETC \fR
+If set the location of the directory containing configuration files
+is \fI$HOME_ETC/.elinks/\fR instead of \fI~/.elinks/\fR.
+.TP
 \fB WWW_HOME \fR
 Homepage location (as in \fBlynx(1)\fR)
 .TP
