diff -urN elinks-20020302/lua.c aaa/elinks-20020302/lua.c
--- elinks-20020302/lua.c	Tue Feb 12 13:18:41 2002
+++ aaa/elinks-20020302/lua.c	Sun Mar  3 21:27:56 2002
@@ -468,6 +468,7 @@
 	lua_register(L, "edit_bookmark_dialog", l_edit_bookmark_dialog);
 	lua_register(L, "xdialog", l_xdialog);
 	do_hooks_file(L, "/etc/", "links-hooks.lua");
+	do_hooks_file(L, "/etc/", "elinks-hooks.lua");
 	if (links_home) do_hooks_file(L, links_home, "hooks.lua");
 }
 
