diff -urN aaa/elinks-0.4pre7/contrib/hooks.lua elinks-0.4pre7/contrib/hooks.lua
--- aaa/elinks-0.4pre7/contrib/hooks.lua	Sun May  5 17:10:20 2002
+++ elinks-0.4pre7/contrib/hooks.lua	Wed May 15 12:39:32 2002
@@ -55,6 +55,7 @@
 -- For any other lua script to be loaded, clone following line:
 
     dofile (home_dir.."/.links/script.lua")
+		dofile ("/usr/share/elinks/script.lua")
 
 
 ----------------------------------------------------------------------
@@ -350,6 +351,7 @@
 if bookmark_addon then
 
     dofile (home_dir.."/.links/bm.lua")
+		dofile ("/usr/share/elinks/bm.lua")
 
     -- Add/change any bookmark options here.
 
