--- elinks-0.12pre1/doc/man/man1/elinks.1.in.old	2008-07-01 02:11:44.000000000 +0200
+++ elinks-0.12pre1/doc/man/man1/elinks.1.in	2008-07-01 18:36:50.000000000 +0200
@@ -315,6 +315,12 @@ The path to the users home directory. Us
 ~/.
 .RE
 .PP
+HOME_ETC
+.RS 4
+If set the location of the directory containing configuration files
+is \fI$HOME_ETC/.elinks/\fR instead of \fI~/.elinks/\fR.
+.RE
+.PP
 WWW_HOME
 .RS 4
 Homepage location (as in
--- elinks-0.12pre1/src/config/home.c.old	2008-07-01 02:11:44.000000000 +0200
+++ elinks-0.12pre1/src/config/home.c	2008-07-01 18:41:01.000000000 +0200
@@ -110,7 +110,8 @@ static unsigned char *
 get_home(void)
 {
 	unsigned char *home_elinks;
-	unsigned char *envhome = getenv("HOME");
+	unsigned char *home_etc = getenv("HOME_ETC");
+	unsigned char *envhome = home_etc ? home_etc : getenv("HOME");
 	unsigned char *home = NULL;
 
 	if (!home && envhome)
