--- elinks-0.9.1/configure.in~	Mon Jul 26 00:42:36 2004
+++ elinks-0.9.1/configure.in	Mon Jul 26 00:42:36 2004
@@ -793,7 +793,7 @@
 dnl Gettext grey zone. Beware.
 dnl ===================================================================
 
-ALL_LINGUAS="be bg ca cs da de el es et fi fr gl hr hu id is it leet lt nl no pl pt pt_BR ro ru sk sv tr uk"
+ALL_LINGUAS="be bg ca cs da de el es et fi fr gl hr hu id is it leet lt nb nl pl pt pt_BR ro ru sk sv tr uk"
 
 AM_GNU_GETTEXT
 
--- elinks-0.9.1/configure~	Mon Jul 26 00:42:28 2004
+++ elinks-0.9.1/configure	Mon Jul 26 00:42:28 2004
@@ -5633,7 +5633,7 @@
 
 
 
-ALL_LINGUAS="be bg ca cs da de el es et fi fr gl hr hu id is it leet lt nl no pl pt pt_BR ro ru sk sv tr uk"
+ALL_LINGUAS="be bg ca cs da de el es et fi fr gl hr hu id is it leet lt nb nl pl pt pt_BR ro ru sk sv tr uk"
 
 
 # Make sure we can run config.sub.
