--- elinks-0.5pre10/src/scripting/lua/core.c.orig	2003-11-14 02:31:06.000000000 +0000
+++ elinks-0.5pre10/src/scripting/lua/core.c	2003-12-02 11:41:05.000000000 +0000
@@ -521,10 +521,12 @@
 	if (elinks_home) do_hooks_file(L, elinks_home, "hooks.lua");
 }
 
+static void free_lua_console_history_entries(void);
+	
 static void
 cleanup_lua(struct module *module)
 {
-	free_lua_console_history(NULL, NULL);
+	free_lua_console_history_entries();
 	lua_close(L);
 }
 
@@ -703,10 +705,16 @@
 	return EHS_NEXT;
 }
 
+static void
+free_lua_console_history_entries(void)
+{
+	free_list(lua_console_history.entries);
+}
+
 enum evhook_status
 free_lua_console_history(va_list ap, void *data)
 {
-	free_list(lua_console_history.entries);
+	free_lua_console_history_entries();
 	return EHS_NEXT;
 }
 
