--- elinks-0.11.3/src/util/file.c.old	2007-06-19 09:12:03.000000000 +0200
+++ elinks-0.11.3/src/util/file.c	2007-06-19 09:12:47.000000000 +0200
@@ -461,16 +461,16 @@
 
 		if (current_time > when + 6L * 30L * 24L * 60L * 60L
 		    || current_time < when - 60L * 60L)
-			fmt = "%b %e  %Y";
+			fmt = "%4b %e  %Y";
 		else
-			fmt = "%b %e %H:%M";
+			fmt = "%4b %e %H:%M";
 
 		add_date_to_string(string, fmt, &when);
 		add_char_to_string(string, ' ');
 		return;
 	}
 #endif
-	add_to_string(string, "             ");
+	add_to_string(string, "              ");
 }
 
 
--- elinks-0.11.3/src/protocol/ftp/ftp.c.old	2007-06-19 09:13:09.000000000 +0200
+++ elinks-0.11.3/src/protocol/ftp/ftp.c	2007-06-19 09:13:52.000000000 +0200
@@ -1157,9 +1157,9 @@
 
 		if (current_time > when + 6L * 30L * 24L * 60L * 60L
 		    || current_time < when - 60L * 60L)
-			fmt = "%b %e  %Y";
+			fmt = "%4b %e  %Y";
 		else
-			fmt = "%b %e %H:%M";
+			fmt = "%4b %e %H:%M";
 
 		wr = strftime(date, sizeof(date), fmt, when_tm);
 
@@ -1168,7 +1168,7 @@
 		add_to_string(&string, date);
 	} else
 #endif
-	add_to_string(&string, "            ");
+	add_to_string(&string, "             ");
 
 	add_char_to_string(&string, ' ');
 
