--- elinks-0.4pre7/configure.in.x	Sat May 11 22:25:38 2002
+++ elinks-0.4pre7/configure.in	Sat May 11 22:23:55 2002
@@ -227,7 +227,7 @@
 	for luadir in "" /usr /usr/local; do
 		if test "$cf_result" = no; then
 			if test ! -z "$luadir"; then
-				LIBS="-L$luadir/lib -llua -llualib -lm $LIBS_X"
+				LIBS="-L$luadir/lib -llua -llualib -lm -ldl $LIBS_X"
 				CFLAGS="$CFLAGS_X -I$luadir/include -I$luadir/include/lua"
 			else
 				LIBS="-llua -llualib -lm $LIBS_X"
