diff -urNp -x '*.orig' elinks-0.13-20180901.org/doc/man/man1/elinks.1.in elinks-0.13-20180901/doc/man/man1/elinks.1.in
--- elinks-0.13-20180901.org/doc/man/man1/elinks.1.in	2017-07-23 19:50:15.000000000 +0200
+++ elinks-0.13-20180901/doc/man/man1/elinks.1.in	2021-05-07 22:58:22.368520341 +0200
@@ -322,6 +322,12 @@ The path to the users home directory\&.
 ~/\&.
 .RE
 .PP
+HOME_ETC
+.RS 4
+If set the location of the directory containing configuration files
+is \fI$HOME_ETC/.elinks/\fR instead of \fI~/.elinks/\fR.
+.RE
+.PP
 WWW_HOME
 .RS 4
 Homepage location (as in
diff -urNp -x '*.orig' elinks-0.13-20180901.org/src/config/home.c elinks-0.13-20180901/src/config/home.c
--- elinks-0.13-20180901.org/src/config/home.c	2017-07-23 19:50:15.000000000 +0200
+++ elinks-0.13-20180901/src/config/home.c	2021-05-07 22:58:22.368520341 +0200
@@ -110,7 +110,8 @@ static unsigned char *
 get_home(void)
 {
 	unsigned char *home_elinks;
-	unsigned char *envhome = getenv("HOME");
+	unsigned char *home_etc = getenv("HOME_ETC");
+	unsigned char *envhome = home_etc ? home_etc : getenv("HOME");
 	unsigned char *home = NULL;
 
 	if (!home && envhome)
